gtn/functions/compose.cpp

```diff
@@ -450,3 +451,665 @@ Graph compose(
 
 } // namespace detail
 } // namespace gtn
+
+namespace gtn {
+namespace detail {
+namespace dataparallel {
+// Change AOS to SOA
+struct GraphDataParallel {
+  std::vector<int> accept;
+  std::vector<int> start;
+
+  // Has one more element than the number of nodes
+  // One value per node - i-th value corresponds to i-th node
+  // Last element is the total number of arcs, so that
+  // each element and its neighbor forms a range 
+  std::vector<int> inArcOffset;
+  std::vector<int> outArcOffset;
+
+  // One value per arc
+  std::vector<int> inArcs;
+  std::vector<int> outArcs;
+
+  // One value per arc
+  // i-th value corresponds to i-th arc
+  std::vector<int> ilabels;
+  std::vector<int> olabels;
+  std::vector<int> srcNodes;
+  std::vector<int> dstNodes;
+  std::vector<float> weights;
+};
+
+namespace {
+// Exclusive/Inclusive prefix sum. The returned vector
+// has one more element 
+void prefixSumScan(std::vector<int>& input) {
+  int sum = 0;
+  for (auto i : input.size()) {
+    auto count = input[i];
+    input[i] = sum;
+    sum += count;
+  }
+  input.push_back(sum);
+}
+
+// Removes all elements that match toRemove from a vector
+std::vector<int> streamCompact(const std::vector<int>& input, int toRemove) {
+  std::vector<int> output;
+
+  for (auto i : input.size()) {
+    if (input[i] != toRemove) {
+      output.push_back(input[i]);
+    }
+  }
+
+  return output;
+}
+
+// TODO: Duplicate - should be removed
+inline int TwoDToOneDIndex(int n1, int n2, int n1Extent) {
+  return n1 + n2 * n1Extent;
+}
+
+inline std::pair<int, int> OneDToTwoDIndex(int n, int n1Extent) {
+  assert(n1Extent > 0);
+  const int n2 = n / n1Extent;
+  const int n1 = n % n1Extent;
+  return std::make_pair(n1, n2);
+}
+
+inline bool checkAnyTrue(const std::vector<bool>& flags) {
+  for (auto i : flags) {
+    if (i == true) {
+      return i;
+    }
+  }
+  return false;
+}
+
+// Map thread id to corresponding node and edge pair
+// Search to find which node pair this tid will fall into
+// Linear search for now (edgeMullOffset is sorted by definition)
+std::pair<std::pair<int, int>, std::pair<int, int>> computeNodeAndEdgePair(
+    int tid,
+    const std::vector<int>& edgeMullOffset,
+    const std::vector<std::pair<int, int>>& toExploreNodePair) {
+  std::pair<int, int> nodePair;
+  std::pair<int, int> edgePair;
+  for (size_t i = 0; i < edgeMullOffset.size() - 1; ++i) {
```

Comment:
If `edgeMullOffset` is empty, `edgeMullOffset.size() - 1` is a very large
number in terms of unsigned integers.

It's error-prone to compare with unsigned integers.

Reply:
I would have compared that with `i + 1 < edgeMullOffset.size()`.